lib/granite/rubocop/cop/performer_in_initializer.rb

```diff
@@ -0,0 +1,198 @@
+require 'rubocop'
+
+module RuboCop
+  module Cop
+    module Granite
+
+      # Checks if performer is being passed to the BA initializer.
+      #
+      # Instead of initializing, use `.as(performer)` or `.as_system`
+      #
+      # @example
+      #   # bad
+      #   BA::Subject::Action.new(performer: Role.system, other: 'arg')
+      #
+      #   # good
+      #   BA::Subject::Action.as_system.new(other: 'arg')
+      #
+      #   # bad
+      #   BA::Subject::Action.new(subject, performer: subject.author, other: 'arg')
+      #
+      #   # good
+      #   BA::Subject::Action.as(subject.author).new(subject, other: 'arg')
+      class PerformerInInitializer < ::RuboCop::Cop::Cop
+        MSG = 'Use `.as(performer)` or `.as_system` instead of passing performer to the initializer'.freeze
+
+        def on_send(node)
+          return unless business_action?(node) || testing_ba?(node)
+
+          add_offense(node, location: :expression, message: MSG)
+        end
+
+        def autocorrect(node)
+          return if has_method_calls_before_new?(node)
+
+          node = described_class_initialized?(node).parent if testing_ba?(node)
+          if has_performer_as_role?(node)
+            autocorrect_performer_as_role_system(node)
+          else
+            autocorrect_performer_as_expression(node)
+          end
+        end
+
+        private
+
+        def_node_search :under_ba_namespace?, '(:const nil? :BA)'
+        def_node_matcher :described_class_initialized?, '^(send (send _ :described_class) :new $_)'
+        def_node_matcher :describe_block?, '(block (send _ :describe $...) ...)'
+        def_node_search :has_performer_key_in_params?, '(sym :performer)'
+        def_node_search :has_as_expression?, ' (send  _  :as _) '
+        def_node_search :has_performer_as_role?, '(pair #has_performer_key_in_params? (send _ :system))'
+        def_node_search :has_performer_as_role, '(pair #has_performer_key_in_params? (send _ :system))'
+        def_node_search :has_performer_as_expression, '(pair #has_performer_key_in_params? $_)'
+        def_node_matcher :has_method_calls_before_new?, '(send (send ...) :new ...)'
+
+        def business_action?(node)
+          return unless initializer?(node)
+          return unless has_performer_key_in_params?(node)
+
+          under_ba_namespace?(node) && !already_fixed?(node)
+        end
+
+        def already_fixed?(node)
+          has_as_expression?(node)
+        end
+
+        def testing_ba?(node)
+          return false unless described_class_is_ba?(node)
+
+          params = described_class_initialized?(node)
+          params && has_performer_key_in_params?(params)
+        end
+
+        def described_class_is_ba?(node)
+          parent = node.parent
+          while parent
+            describing_class = describe_block?(parent)
+            if describing_class
+              return false if describing_class.empty?
```

Comment:
Is `.empty?` necessary?

Reply:
According to specs it is. I'll leave it as is, feel free to unresolve this issue if you insist on me investigating :+1: